### PR TITLE
Allow unsigned integers in switch

### LINKED ISSUE #1029: [wgsl-in] Naga complains about unsigned literal in switch statement
Naga (1f42d4f) rejects this WGSL (which was generated by Tint, via the example from issue #1028):

```
[[stage(fragment)]]
fn main() {
  switch(0u) {
    default: {
      if (false) {
        break;
      }
    }
  }
  return;
}
```

I get this output:

```
$ RUST_BACKTRACE=1 cargo run temp.wgsl blah.spv
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
     Running `target/debug/naga temp.wgsl blah.spv`
Entry point main at Fragment is invalid: 
	The `switch` value [1] is not an integer scalar
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', cli/src/main.rs:272:59
stack backtrace:
   0: rust_begin_unwind
             at /rustc/24bdc6d73a75dce9a7013ebc7c037013ff4ea099/library/std/src/panicking.rs:515:5
   1: core::panicking::panic_fmt
             at /rustc/24bdc6d73a75dce9a7013ebc7c037013ff4ea099/library/core/src/panicking.rs:92:14
   2: core::panicking::panic
             at /rustc/24bdc6d73a75dce9a7013ebc7c037013ff4ea099/library/core/src/panicking.rs:50:5
   3: core::option::Option<T>::unwrap
             at /rustc/24bdc6d73a75dce9a7013ebc7c037013ff4ea099/library/core/src/option.rs:388:21
   4: naga::main
             at ./cli/src/main.rs:272:45
   5: core::ops::function::FnOnce::call_once
             at /rustc/24bdc6d73a75dce9a7013ebc7c037013ff4ea099/library/core/src/ops/function.rs:227:5
```

### PR DESCRIPTION
**Pull request by JCapucho**
The wgsl spec defines that the switch selector must be a scalar integer type, this encompasses both signed and unsigned integers. This PR adds support for unsigned integers.

Furthermore the spec defines that the type of the selector and the case value type must match this PR also adds validation for that.

Closes #1029

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1251,15 +1251,24 @@ pub enum Expression {
...
 /// A case for a switch statement.
 // Clone is used only for error reporting and is not intended for end users
 #[derive(Clone, Debug)]
 #[cfg_attr(feature = "serialize", derive(Serialize))]
 #[cfg_attr(feature = "deserialize", derive(Deserialize))]
 pub struct SwitchCase {
     /// Value, upon which the case is considered true.
-    pub value: i32,
-    /// Body of the cae.
+    pub value: SwitchCaseValue,
```
Comment (kvark) [inline src/lib.rs:1270]: is this really needed?
  1. different cases inside the same switch must have the same type
  2. both i32 and u32 can be losslessly converted to each other

So it seems to me that we could just keep `i32` (or `u32`) in here, and just interpret it differently based on some bit inside the switch itself.

**Review by kvark — CHANGES_REQUESTED** [review]
Got a few questions about the IR

File: src/lib.rs (hunk 1 above)
Comment (JCapucho) [reply src/lib.rs:1270]: My fear is how you would check that they have the same type?

File: src/back/mod.rs [hunk 2]
```diff
@@ -311,3 +311,12 @@ impl crate::TypeInner {
         }
     }
 }
+
+impl std::fmt::Display for crate::SwitchCaseValue {
```
[the remark this answers is not in the document]
Comment (JCapucho) [reply src/back/mod.rs:315]: I can move the implementation elsewhere but it's useful since all backends use the C-like syntax for their integer literals, except spirv which is a binary representation.

File: src/lib.rs (hunk 1 above)
Comment (kvark) [reply src/lib.rs:1270]: WGSL frontend can check this. The IR would just assume they are the same type

Comment (JCapucho) [reply src/lib.rs:1270]: And the backends would need to check the type of the selector and write the postfix as needed?

Comment (kvark) [reply src/lib.rs:1270]: correct, yes

**Comment by JCapucho** [comment]
@kvark updated to always use `i32`

File: src/back/wgsl/writer.rs [hunk 3]
```diff
@@ -880,13 +880,22 @@ impl<W: Write> Writer<W> {
...
+                    .scalar_kind()
+                    == Some(crate::ScalarKind::Uint);
                 writeln!(self.out, ") {{")?;
 
                 // Write all cases
                 let mut write_case = true;
                 let all_fall_through = cases
                     .iter()
                     .all(|case| case.fall_through && case.body.is_empty());
+                let postfix = || match uint {
```
Comment (kvark) [inline src/back/wgsl/writer.rs:895]: using closures here seems like an overkill. Couldn't we just do:
```rust
let type_postfix = match func_ctx.info[selector].ty.inner_with(&module.types) {
  crate::TypeInner::Scalar { kind: crate::ScalarKind::Uint, .. } => "u",
  _ => "",
};
```
I think doing this in other places would also improve things a little.

**Review by kvark — COMMENTED** [review]
Thank you! just one last thing

File: src/back/wgsl/writer.rs (hunk 3 above)
Comment (JCapucho) [reply src/back/wgsl/writer.rs:895]: ok I realized what you meant now

File: tests/in/control-flow.wgsl [hunk 4]
```diff
@@ -32,6 +32,13 @@ fn main([[builtin(global_invocation_id)]] global_id: vec3<u32>) {
         }
     }
 
+	// switch with unsigned integer selectors
+	switch(0u) {
+		case 0u: {
+			break;
```
Comment (kvark) [inline tests/in/control-flow.wgsl:38]: why do we `break` here? We can do it, but it seems irrelevant to the PR.

**Review by kvark — CHANGES_REQUESTED** [review]

**Review by kvark — APPROVED** [review]
